### ISSUE TITLE
[ENH] reducer design for rework

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/sktime#3224]
I think we need a small design document for the rework of the reducers. FYI @danbartl, given recent discussions.

Cases we should cover with modular design:

* lagging `X` and lagging `y`, lagging `y` but not lagging `X`
* use of contemporary `X`, use of shifted `X` (by forecasting horizon)
* use of transformers on `X`, e.g., window summarizer
* removing missing values created due to shifting/windowing
* treating missing values created due to shifting/windowing (e.g. filling)
* treating missing values from shifting/windowing differently from pre-existing missing values, including options to leave the latter be

**Comment by ltsaprounis** [comment]
One more point from me:
- Use of transformers that are applied on the index such as the `DateTimeFeatures` transformer and other transformers that could be applied to the hierarchy features of hierarchical time series.

**Comment by ltsaprounis** [comment]
> * use of contemporary `X`, use of shifted `X` (by forecasting horizon)
Would that force passing `fh` on fit?

**Comment by fkiraly** [comment]
> Would that force passing `fh` on fit?

I suppose so?

**Comment by fkiraly** [comment]
PR with a prototype design for direct and recursive forecasting:
https://github.com/alan-turing-institute/sktime/pull/3333

Solves the above points as follows:

>lagging X and lagging y, lagging y but not lagging X

reducers by default *do not apply any transformations (incl lagging) to `X`*, only to `y`.
* to lag `X`, the `Lag` transformer needs to be applied in a pipeline to exogeneous data.
* to lag `X` and not `y`, apply the `Lag` trafo to exog, and leave the window length parameter at zero.
* lagging `y` an not `X` is the standard behaviour.

 > use of contemporary X, use of shifted X (by forecasting horizon)

In direct reduction, these are options, via parameter `X_treatment` that can be `"concurrent"` or `"shifted"`. In recursive reduction, concurrent is default. Shifted is obtained by lagging `X` with a lag of 1 (we could turn thsi into a parameter, too).

> use of transformers on X, e.g., window summarizer

Transformers for `X` are not passed to the reducers, but need to be applied via a pipeline.
This is more modular, and avoids duplication and "god class" like parameter handling.

> removing missing values created due to shifting/windowing

done by parameter `impute` which is looped through to `Imputer`.
All options in `Imputer` are available in reduction.

`Imputer` has no missing value removal, but that could be added as an option, and would also add the option to the reducers, via the estimator modular dependency.

> treating missing values created due to shifting/windowing (e.g. filling)

option via the `Imputer`.

> treating missing values from shifting/windowing differently from pre-existing missing values, including options to leave the latter be

This is solved by the possibility to attach another `Imputer` earlier in the pipeline, this handles only pre-existing missing values. The missing values caused by the reduction are handled by the `impute` setting, and the strategies for the two could in-principle be different.

**Comment by fkiraly** [comment]
FYI @KishManani, @ngupta23, @wregter, @danbartl, @ltsaprounis, @grzegorzrut - any comments?

**Comment by ilkersigirci** [comment]
Thank you for this refactor. ` RecursiveReductionForecaster`  doesn't work with keras models that are imported from `sktime-dl` unlike `make_reduction` scitype=time-series-regressor, right? It seems to just work with sklearn models

**Comment by wregter** [comment]
These changes are great. I was wondering: what about the use of covariates in X of which we do not know the future values? Sklearn regressors are capable of handling these, but might they need any special treatment in the reducer? For example, when we have a dataset of past data we do have the future (test) values for these covariates. But in order to test a realistic scenario, we would need to set our test set's future values for this covariate to NA (I assume). Could this perhaps be something handled by the reducer? Or maybe in the form of a transformer, that can be passed to the reducer?

**Comment by ilkersigirci** [comment]
In the new design, `_predict_out_of_sample` takes approximately 10x more time than old design. After some debugging, I have found out that [these fit_transform](https://github.com/alan-turing-institute/sktime/blob/26b8146ae4ddec66d82c9db533a59abd33503de4/sktime/forecasting/compose/_reduce.py#L1889) codes are the main culprit. FYI @fkiraly

**Comment by fkiraly** [comment]
> Thank you for this refactor. ` RecursiveReductionForecaster`  doesn't work with keras models that are imported from `sktime-dl` unlike `make_reduction` scitype=time-series-regressor, right? It seems to just work with sklearn models 

@ilkersigirci, correct - the current prototypes do not work with time series regressors (yet), but that's an easy change down the line. This is more for aligning on the design, e.g., are the parameters the right ones, is this intuitive to use, etc.

**Comment by fkiraly** [comment]
> In the new design, `_predict_out_of_sample` takes approximately 10x more time than old design. After some debugging, I have found out that [these fit_transform](https://github.com/alan-turing-institute/sktime/blob/26b8146ae4ddec66d82c9db533a59abd33503de4/sktime/forecasting/compose/_reduce.py#L1889) codes are the main culprit. FYI @fkiraly 

@ilkersigirci, really? That's unexpected. Is this not just a wrapper of pandas shift?

Do you know whether the time is lost in conversion, e.g., having a numpy implementation for `Lag` would solve it? (I can then get to it, it's easy)

**Comment by fkiraly** [comment]
> These changes are great. I was wondering: what about the use of covariates in X of which we do not know the future values? 

There are three ways I can think of:
* use `"shifted"` mode in the direct reducer, that requires only past values of covariates. In this mode, we are shifting the past values of `X` to align with the future values of `y`, so no NA are created.
* use a transformer that moves past `X` to the future and/or transforms `X`, in a `**` pipeline/`ForecastingPipeline`
* use the `ForecastX` wrapper

**Comment by ilkersigirci** [comment]
> > In the new design, `_predict_out_of_sample` takes approximately 10x more time than old design. After some debugging, I have found out that [these fit_transform](https://github.com/alan-turing-institute/sktime/blob/26b8146ae4ddec66d82c9db533a59abd33503de4/sktime/forecasting/compose/_reduce.py#L1889) codes are the main culprit. FYI @fkiraly 
> 
> @ilkersigirci, really? That's unexpected. Is this not just a wrapper of pandas shift?
> 
> Do you know whether the time is lost in conversion, e.g., having a numpy implementation for `Lag` would solve it? (I can then get to it, it's easy)

When I analyze the code with `profile` in pycharm to see which function has taken too much time, I have found that execution of numpy diff function took most of the time. I don't know if simple conversion to numpy fix this problem but I think this is a crucial point since for example forecasting 30 pointz ahead takes 6 seconds with lightgbm. It should be nearly instant.

**Comment by fkiraly** [comment]
> Thank you for your proposition @fkiraly. Isn't it more general to use a dictionary to define the windows and functions to apply to `y` or to `X` as in `WindowSummarizer` ? This way we can define the columns to lags in X, when the lag starts, which aggregate window to apply; similarly for `y` without the choice of the column. 

I thought of this - we could simply patch a collection of lag through to the `Lag` transformer used internally, here's from the docstring on the `lags` value:

```
lags : lag offset, or list of lag offsets, optional, default=0 (identity transform)
        a "lag offset" can be one of the following:
        int - number of periods to shift/lag
        time-like: DateOffset, tseries.offsets, or timedelta
            time delta offset to shift/lag
            requires time index of transformed data to be time-like (not int)
        str - time rule from pandas.tseries module, e.g., "EOM"
```

Though I thought, it might be too confusing?
Since a window is a common choice, so passing an integer would hav ambiguous intuitive meaning, i.e., "7" could mean: only the lag of 7, vs all lags 1, 2, ... 7.

On the other hand, configurability might be helpful.

Ideas/thoughts on this? What would you personally prefer?
* the same parameter as `Lag` has?
* only int is possible, and that indicates a full window (as current)
* two different parameters, e.g., `lags` and `window_length`, and only one of these can be passed. `window_length` behaves as 2nd bullet point, `lags` behaves as the arg of `Lag`

**Comment by fkiraly** [comment]
> This way we can define the columns to lags in X, when the lag starts, which aggregate window to apply; similarly for y without the choice of the column.

we could also allow passing a transformer that is applied to `y`? E.g., `WindowSummarizer`.

**Comment by ltsaprounis** [comment]
> These changes are great. I was wondering: what about the use of covariates in X of which we do not know the future values?

I think the clean solution in this case is adding a new variable, say, Z, that represents these observed values that are not the target and we don't have their future values. The solution of shifting those based on the horizon is a hack to avoid this.

**Comment by ltsaprounis** [comment]
Also we need a way to create exogenous variables based on the absolute horizon. 
For example, dummy seasonality ([DateTimeFeatures](https://github.com/alan-turing-institute/sktime/blob/0bd87682f0b67ff2fec49921a43d2fc17674205a/sktime/transformations/series/date.py)) and/or Fourier features (coming soon) are exogenous variables of the form of X that are generated from the time index only.

In general, I think that reduction is comprised of 4 feature engineering tasks (or 5 if we want to count static covariates): 

1. Create features from the target, y
2. Create features from the time index (e.g. DateTimeFeatures) 
3. Create features from exogenous variables where their future values are known, X
4. Create features from exogenous variables where their future values are NOT known, Z
5. Create features from static covariates (e.g. label encode the index of the hierarchy levels or the unique instances etc.)

So the following composite estimator would do the job:
```python
class CompositeReducer(BaseForecaster):
"""Reduction through feature engineering pipelines applied on y, X, Z and the time index

    Parameters
    ----------
    y_transformer: BaseTransformer, default = None
        Feature engineering transformer for y. e.g. Lag(5) or WindowSummarizer()
    X_transformer: BaseTransformer, optional, default = None
        Feature engineering transformer for X. e.g. Lag(5) or WindowSummarizer(target_cols=[""col_1", col_3"])
    Z_transformer: BaseTransformer, optional, default = None
        Feature engineering transformer for Y. e.g. Lag(5) or WindowSummarizer(target_cols=[""col_1", col_3"])
    time_index_transformer: BaseTransformer, default = None
        Feature engineering transformer for time index e.g. DateTimeFeatures()
    ...
"""
```

What do you think?

**Comment by fkiraly** [comment]
>> These changes are great. I was wondering: what about the use of covariates in X of which we do not know the future values?

> I think the clean solution in this case is adding a new variable, say, Z, that represents these observed values that are not the target and we don't have their future values. The solution of shifting those based on the horizon is a hack to avoid this.

Why does adding a new variable in the interface solve this? You would still have to shift them if you want to use contemporaneous direct reduction.

**Comment by fkiraly** [comment]
>>> In the new design, `_predict_out_of_sample` takes approximately 10x more time than old design. After some debugging, I have found out that [these fit_transform](https://github.com/alan-turing-institute/sktime/blob/26b8146ae4ddec66d82c9db533a59abd33503de4/sktime/forecasting/compose/_reduce.py#L1889) codes are the main culprit. FYI @fkiraly 
>> 
>> @ilkersigirci, really? That's unexpected. Is this not just a wrapper of pandas shift?
>> 
>> Do you know whether the time is lost in conversion, e.g., having a numpy implementation for `Lag` would solve it? (I can then get to it, it's easy)

> When I analyze the code with `profile` in pycharm to see which function has taken too much time, I have found that execution of numpy diff function took most of the time. I don't know if simple conversion to numpy fix this problem but I think this is a crucial point since for example forecasting 30 pointz ahead takes 6 seconds with lightgbm. It should be nearly instant.

That's odd, because we use `pandas.shift` under the hood. Could you post some minimal self-contained code?

Here's another thing that it could be, if you are using hierarchical data:
the old reducer, by default, fits per instance, and it cannot do global fitting. The new reducer, by default, fits on the entire pooled data.
To compare like with like, set `pooling = "local"` in the new reducers.
Of course, if your data is a single time series, there will be no difference.

**Comment by ilkersigirci** [comment]
I am using single time series data. For example purposes, I experimented with `load_airline` data. As you can see in the code, oos time difference between two methods is huge

```python
from lightgbm import LGBMRegressor
from sktime.forecasting.compose import (
    make_reduction,
    RecursiveReductionForecaster
)
import pandas as pd
import numpy as np
import time

from sktime.datasets import load_airline
from sktime.forecasting.model_selection import temporal_train_test_split

y = load_airline()
X = pd.DataFrame(
    {
        "lag1": y.shift(1).fillna(method="bfill"),
        "lag2": y.shift(2).fillna(method="bfill"),
    },
    index=y.index,
)

y_train, y_test, X_train, X_test = temporal_train_test_split(y=y, X=X, test_size=30)
fh = np.arange(1, len(y_test) + 1)


old_model = make_reduction(
    estimator=LGBMRegressor(),
    scitype="tabular-regressor",
    window_length=1,
    strategy="recursive",
)

new_model = RecursiveReductionForecaster(
    estimator=LGBMRegressor(),
    window_length=1,
    # impute_method="bfill",
    impute_method=None,
    pooling="local",
)

old_model.fit(y=y_train, X=X_train, fh=fh)

old_model_start_forecast = time.time()
old_model.predict(X=X_test, fh=fh)
old_model_end_forecast = time.time()
print(f"Old model forecast time: {old_model_end_forecast - old_model_start_forecast}") # 0.009313583374023438

new_model.fit(y=y_train, X=X_train)

new_model_start_forecast = time.time()
new_model.predict(X=X_test, fh=fh)
new_model_end_forecast = time.time()
print(f"New model forecast time: {new_model_end_forecast - new_model_start_forecast}") # 1.0612244606018066

```

**Comment by fkiraly** [comment]
possible solution to no.3, no. 4 - allowing transformers that operate on the resulting tabular data frame before passed to the sklearn transformer

**Comment by fkiraly** [comment]
from @LoicRaillon and @ltsaprounis both: we should be able to specify the lags for `y`, they do not need to be all consecutive lags

**Comment by ilkersigirci** [comment]
@fkiraly I would like to add one more point to this refactor. Both old and the new code concat given `y`(lagged form of it) with `X`(if exists) , to create new `X` for training sklearn estimator. This is useful if we don't have any `X` values since we must somewhat convert our data to tabular format. But, if `X` exists, meaning we already have a tabular formed data,  I think we should have an option to **not** to  concat `y` values to `X`. What do you think about this? I don't want the sklearn models to always see the lagged `y` values as a feature if I feed them with `X` explicitly.

**Comment by KishManani** [comment]
@fkiraly - some of my thoughts.

I agree with @ilkersigirci that it would be helpful to not be forced to use lag features. It would be nice to have the option to use other transformations of the target time series (e.g., a lagged rolling window) instead of being forced to use a lag feature on top of other transformations we apply to the target. 

Also, I think point 5 from @ltsaprounis' [comment](https://github.com/alan-turing-institute/sktime/issues/3224#issuecomment-1227669136) would be very helpful. Addressing point 5 may mean allowing the flexibility of being able to pass a transformer that has access to the full dataset and being able to create features from the target and other variables (e.g., target encoding a static feature as @ltsaprounis pointed out).

**Comment by fkiraly** [comment]
> @fkiraly I would like to add one more point to this refactor. Both old and the new code concat given y(lagged form of it) with X(if exists) , to create new X for training sklearn estimator. This is useful if we don't have any X values since we must somewhat convert our data to tabular format. But, if X exists, meaning we already have a tabular formed data, I think we should have an option to not to concat y values to X. What do you think about this? I don't want the sklearn models to always see the lagged y values as a feature if I feed them with X explicitly.

@ilkersigirci, in the prototype, that's simply the default option, window length equals zero.
(I think?)

**Comment by fkiraly** [comment]
@KishManani, @ltsaprounis, re static covariates, good point!
How should these be passed to the models, though?

Only one of the data types (`nested_univ`) allows them currently, and it is not used by most estimators.

**Comment by ilkersigirci** [comment]
> > @fkiraly I would like to add one more point to this refactor. Both old and the new code concat given y(lagged form of it) with X(if exists) , to create new X for training sklearn estimator. This is useful if we don't have any X values since we must somewhat convert our data to tabular format. But, if X exists, meaning we already have a tabular formed data, I think we should have an option to not to concat y values to X. What do you think about this? I don't want the sklearn models to always see the lagged y values as a feature if I feed them with X explicitly.
> 
> @ilkersigirci, @KishManani, in the prototype, that's simply the default option, window length equals zero. (I think?)

`window_length=0` gives `ValueError: No objects to concatenate`. Hence I have done my tests with `window_length=1`

**Comment by fkiraly** [comment]
> The ValueError doesn't happen on your side @fkiraly ? I am using the latest code that you committed

Yes, you are right, @ilkersigirci.
The `window_length=0` behaviour was only in my head & written down in a tentative design, not in the actual code!

**Comment by fkiraly** [comment]
say, do you have any good leads where the slowness is coming from in the prototype, @ilkersigirci?

**Comment by ngupta23** [comment]
Hi @fkiraly and team, Sorry for the delayed response. I was tied up for a while. I would like to see if the following features can be added to this reducer (if not available already)

1. Ability to extract the reduced tabular dataset from the original data.
2. Make in-sample predictions
3. Ability to adjust the in-sample reduced tabular dataset values for exogenous variables and make in-sample predictions. This is useful for checking the impact of exogenous variables on the target variables (i.e. What If Scenarios).
4. Getting feature importance from the regression model used in the modeling.

Thanks!

**Comment by ltsaprounis** [comment]
Thanks for the input @ngupta23!
I agree with the in-sample prediction features + if we design this in a nice way it will allow for purged cross validation for time series! https://www.youtube.com/watch?v=2-_2lShioBE&t=2415s

for number 4 - you can do that now by getting to the `estimator` attribute of a reduction forecaster object.

**Comment by fkiraly** [comment]
@ngupta23:

> Ability to extract the reduced tabular dataset from the original data.

In the newest iteration https://github.com/alan-turing-institute/sktime/pull/3382,
this step sits in a transformer that you can call, the `ReducerTransform`. Storing would clutter memory, no? Or is there a deeper reason why you think that should be an option, as opposed to getting it from `transform`?

>Make in-sample predictions

That should work, already.

> Ability to adjust the in-sample reduced tabular dataset values for exogenous variables and make in-sample predictions. This is useful for checking the impact of exogenous variables on the target variables (i.e. What If Scenarios).

I don't think this should be part of the reducer. You can do this though by using the `ColumnSelect` transformer (or the square brackets dunder) as one of the transformers or in a pipeline with one of the transformers.

> Getting feature importance from the regression model used in the modeling.

You can do that by accessing the wrapped sklearn estimator, if it has feature importance.

**Comment by ltsaprounis** [comment]
> I don't think this should be part of the reducer. You can do this though by using the `ColumnSelect` transformer (or the square brackets dunder) as one of the transformers or in a pipeline with one of the transformers.

Thanks @fkiraly - Can you explain in a bit more detail how this would work? I don't get it.
This case is about predicting an absolute "horizon" in the past with different X than the one passed on train.
e.g. What would have happened to sales of suns-cream on weeks a, b and c if the temperature was 25 degrees instead of 30 degrees which was the actual temp on weeks a, b and c.

**Comment by fkiraly** [comment]
> Can you explain in a bit more detail how this would work? I don't get it.

Oh, I see, I misunderstood. I thought you wanted to select *presence of variables* (present yes/no), but I now understand that you want to switch out *values*.

I suppose this is already possible in all vanilla versions of estimators, as you can pass exogeneous data `X` in `predict` that need not agree with the exogeneous data already seen. If you do that in-sample, you should get what you want? Not just for the reducer, but for any forecasting estimator.

**Comment by ngupta23** [comment]
> > Make in-sample predictions
> 
> That should work, already.

@fkiraly This does not work in the released version of `sktime` today.

![image](https://user-images.githubusercontent.com/33585645/190121216-b9951357-faef-4489-bac7-01b667dad85e.png)

![image](https://user-images.githubusercontent.com/33585645/190121056-0d581cce-34f7-45b5-973d-9837431a0b6a.png)

**Comment by fkiraly** [comment]
> @fkiraly Not sure this works today. This seems to have no impact for the ARIMA model for example. **See [here](http://nbviewer.org/gist/ngupta23/cd6641b830469bf3cb1ccd5e7014d71c)**. The in-sample predictions are the same irrespective of the exogenous variables passed.

Hmmm, odd - might be coming from `pmdarima`? How about other models?

> @fkiraly This does not work in the released version of sktime today.

Which forecaster is this, can you post some self-contained code so we can use it for test-driven development?

**Comment by fkiraly** [comment]
opened issue for the specific topic of in-sample predictions with new exogeneous data, that is an important topic, but it looks like a different issue since it is not just reducer specific: #3432

**Comment by ngupta23** [comment]
> > @fkiraly This does not work in the released version of sktime today.
> 
> Which forecaster is this, can you post some self-contained code so we can use it for test-driven development?

Hi @fkiraly I have updated the notebook to include a minimal example (**[here](http://nbviewer.org/gist/ngupta23/cd6641b830469bf3cb1ccd5e7014d71c)**)

Also, here 

```python
from sktime.datasets import load_airline
from sktime.forecasting.compose import make_reduction
from sklearn.ensemble import GradientBoostingRegressor

y = load_airline()
X = y/2
X.name = "dummy_exogenous"

regressor = GradientBoostingRegressor()
forecaster = make_reduction(regressor, window_length=15, strategy="recursive")
forecaster.fit(y, X)

# Insample predictions with original X ----
y_pred = forecaster.predict(fh=-np.arange(0, len(y)), X=X) 
```

**Comment by fkiraly** [comment]
Re-trying a simplified version of @ilkersigirci's example with new/old reducer:

```python
from sklearn.linear_model import LinearRegression
from sktime.forecasting.compose import make_reduction
from sktime.forecasting.compose._reduce import RecursiveReductionForecaster
import pandas as pd
import numpy as np
import time

from sktime.datasets import load_airline
from sktime.forecasting.model_selection import temporal_train_test_split

y = _make_hierarchical(hierarchy_levels=(2, 10), max_timepoints=30, min_timepoints=30)

y_train, y_test = temporal_train_test_split(y=y, test_size=5)
fh = np.arange(1, len(y_test) + 1)

old_model = make_reduction(
    estimator=LinearRegression(),
    window_length=5,
    strategy="recursive",
    pooling="local",
)

new_model = RecursiveReductionForecaster(
    estimator=LinearRegression(),
    window_length=5,
    impute_method=None,
    pooling="local",
)

old_model.fit(y=y_train, fh=fh)

old_model_start_forecast = time.time()
old_model.predict(fh=fh)
old_model_end_forecast = time.time()
print(f"Old model forecast time: {old_model_end_forecast - old_model_start_forecast}")

new_model.fit(y=y_train)

new_model_start_forecast = time.time()
new_model.predict(fh=fh)
new_model_end_forecast = time.time()
print(f"New model forecast time: {new_model_end_forecast - new_model_start_forecast}")
```

The discrepancy is quite extreme - I still blame input/output checks, but I thought we had taken care of those?

**Comment by danbartl** [comment]
Hi the fixes to input / output checkes definitely work, greatly reducing the calculation times in some of the scripts (especially hierarchical recon) I use. 

However, the fixes were mostly focused on bigger groups, where previously run times increased exponentially with the number of groups. The fixes address this issue.

I assume something else is at fault here.

**Comment by AliWaheed** [comment]
I found a potential issue with `make_reduction()`

Below self-contained example gives the same MAPE whether we include exog variable or not. Is this intended behavior similar to AutoETS which also ignores exog variables?

`import numpy as np
from sktime.datasets import load_airline
from sktime.forecasting.compose import make_reduction
from sktime.performance_metrics.forecasting import \
    mean_absolute_percentage_error as mape

from lightgbm import LGBMRegressor

y = load_airline()

y_train = y.iloc[:-12]
y_test = y.iloc[-12:]

X_train = y_train/2
X_test = y_test/2

# Train & Predict with X
regressor = LGBMRegressor()
forecaster = make_reduction(regressor, window_length=15, strategy="recursive")
forecaster.fit(y_train, X_train)
y_pred = forecaster.predict(fh=np.arange(1, len(y_test)+1), X=X_test) 
print(mape(y_pred, y_test))

# Train & Predict without X
regressor = LGBMRegressor()
forecaster = make_reduction(regressor, window_length=15, strategy="recursive")
forecaster.fit(y_train)
y_pred = forecaster.predict(fh=np.arange(1, len(y_test)+1)) 
print(mape(y_pred, y_test))`

**Comment by AliWaheed** [comment]
I found a potential issue with ```make_reduction()```

Below self-contained example gives the same MAPE whether we include exog variable or not. Is this intended behavior similar to AutoETS which also ignores exog variables?

```import numpy as np
from sktime.datasets import load_airline
from sktime.forecasting.compose import make_reduction
from sktime.performance_metrics.forecasting import
mean_absolute_percentage_error as mape

from lightgbm import LGBMRegressor

y = load_airline()

y_train = y.iloc[:-12]
y_test = y.iloc[-12:]

X_train = y_train/2
X_test = y_test/2

Train & Predict with X
regressor = LGBMRegressor()
forecaster = make_reduction(regressor, window_length=15, strategy="recursive")
forecaster.fit(y_train, X_train)
y_pred = forecaster.predict(fh=np.arange(1, len(y_test)+1), X=X_test)
print(mape(y_pred, y_test))

Train & Predict without X
regressor = LGBMRegressor()
forecaster = make_reduction(regressor, window_length=15, strategy="recursive")
forecaster.fit(y_train)
y_pred = forecaster.predict(fh=np.arange(1, len(y_test)+1))
print(mape(y_pred, y_test))```

**Comment by AliWaheed** [comment]
I found a potential issue with ```make_reduction()```

Below self-contained example gives the same MAPE whether we include exog variable or not. Is this intended behavior similar to AutoETS which also ignores exog variables?

```python
import numpy as np
from sktime.datasets import load_airline
from sktime.forecasting.compose import make_reduction
from sktime.performance_metrics.forecasting import
mean_absolute_percentage_error as mape

from lightgbm import LGBMRegressor

y = load_airline()

y_train = y.iloc[:-12]
y_test = y.iloc[-12:]

X_train = y_train/2
X_test = y_test/2

Train & Predict with X
regressor = LGBMRegressor()
forecaster = make_reduction(regressor, window_length=15, strategy="recursive")
forecaster.fit(y_train, X_train)
y_pred = forecaster.predict(fh=np.arange(1, len(y_test)+1), X=X_test)
print(mape(y_pred, y_test))

Train & Predict without X
regressor = LGBMRegressor()
forecaster = make_reduction(regressor, window_length=15, strategy="recursive")
forecaster.fit(y_train)
y_pred = forecaster.predict(fh=np.arange(1, len(y_test)+1))
print(mape(y_pred, y_test))
```

**Comment by fkiraly** [comment]
I think I have understood where the runtime in the recursive reducer goes, also FYI @davidgilbertson.

The runtime comes from the *prediction* method - which is surprising as it should not be the bottleneck (but `fit`).

Upon closer inspection, the problem becomes clear but the fix is not easy as it requires fiddling with `pd.DataFrame`.

The root cause is the loop in `_predict_out_of_sample`, which recreates the lagged `DataFrame` for each element in the horizon. This is done via inefficient calls to `Lag` classes, and two were constructed and executes `fit_transform` per loop call.

In the example of @ilkersigirci, the `fh` is `range(100)`. One `fh` element takes 0.04 secs (on my computer) for one series, so you get 4 sec per series, and thus 4 * 20 = 80 sec since there are 20 series.

It should be possible to fix this by constructing the lagged data frames only once, that would prevent multiplying with number of `fh` elements, so we arrive at 0.8 sec with the example, which is in line with the old implementation (albeit a bit longer).

**Comment by fkiraly** [comment]
Also FYI @davidgilbertson - we could give the two reducers a `feature_names_` fitted parameter, this would be a list of feature names that are passed to the internal estimator, with the sequence in which they appear as columns..

If you need these addressed by the internal estimator parameters, you can get the names via that.

**Comment by davidgilbertson** [comment]
I don't know if this helps, but for my own understanding I made this diagram of the types of variables I need to keep track of when forecasting (the columns represent different types of variables that require different treatment in the predict loop).

![image](https://github.com/sktime/sktime/assets/4443482/9bbea92a-5d37-4415-a019-42d8b97c4cfd)

* **Future-known**: features that are known in advance for the whole forecasting horizon (this includes features that can't be known or forecast, like "volcano erupted nearby")
* **Future-derived**: features that must be derived from the target step-by-step during prediction
* **Future-forecast**: features that must be forecasted, but can be done in one batch before target prediction

In a system with 'future-derived' features, prediction at each timestep must to two things:
* Calculate the 'future-derived' values, based on past values of the target (potentially a mix of real data and predictions, depending on the timestep)
* Predict the target

The tricky thing (I find) for 'future-derived' features is how to define something like "past 7 days sales" elegantly. Ideally I'd have a single piece of code that could be used both to calculate the values for all the training data (performantly) and incrementally 'extend' the values as I loop through the prediction timesteps.

@ltsaprounis this has overlap with your list in this comment: https://github.com/sktime/sktime/issues/3224#issuecomment-1227669136

**Comment by mlisovyi** [comment]
> Imputer has no missing value removal, but that could be added as an option, and would also add the option to the reducers, via the estimator modular dependency.

@fkiraly  That would be a useful option. I was already trying to understand how to get rid of first "incomplete" samples that are generated by the new classes but were not there in the "old" implmentation

**Comment by mlisovyi** [comment]
## Context
We have set up a small experiment in order to understand what the reductions do internally and if it makes sense in for our use-case. In our use-case we want to **build forecaster using lagged `y` values as well as exogenous `X` values**, additionally, **we do _not_ have future values for `X`**.

## Observations

1. :red_circle: **X values are set to 0, if `pooling="global"` in both old reductions**[^1], but is correct in the new direct class and fails in the new recursive class due to missing `X_treatment="shifted"`. 
2. :red_circle: **X values are "looped-back" for the horizon length, if `pooling="local"` in `RecursiveTabularRegressionForecaster`**[^1], but is correct in the new and old direct reductions.
3. :red_circle: **X is lagged if `pooling="local"`**, but you are already aware
4. :yellow_circle: `RecursiveReductionForecaster` throws an error in the `predict` method, as only "concurrent" treatment of X is implemented, which is better than filling X values by zeros, as the old classes seem to be doing.
5. :yellow_circle: There was and remains a difference in the order of feature-and target columns between recursive and direct reductions. Not a big problem, if there is a transformer that can replicate exactly that, e.g. for use in feature-importance extraction using `shap`.
6. :green_circle: the new reductions preserve data type (pandas objects, whereas ) and thus propagate column names to the sklearn estimators, which could in turn store/use those, but you are already aware
7. :green_circle: Lagging of the target values in reduction seems to be without issues in all configurations.

In turn, a compilation of 1-3 means that there is no configuration of reduction type/pooling that would work as a user would expect in the presence of exogenous variables without future values in the old reductions. All of them silently do strange things in this scenario. 
I'm very happy to see that we can make use of the new `DirectReductionForecaster`, which seems to be safe for this use-case (other than the issue with not being able to drop rows with missing lagged targets, as mentioned above, but one could hope that thre fraction of such rows at the start is small and thus effect of those is minor).

[^1]: This is true only if there are no future values in X, otherwise  there is a different behaviour.

## Output (as markdown)
Pooling = local: https://gist.github.com/mlisovyi/f9944778da6d9ee12626413602543629
Pooling = global: https://gist.github.com/mlisovyi/1920742767e08caa0cea5713895bede0

## Environment
 `sktime==0.25.0; scikit-learn 1.3.2; pandas 2.0.3`

## Code
```python
import pandas as pd
from sklearn.dummy import DummyRegressor
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import DirectTabularRegressionForecaster, RecursiveTabularRegressionForecaster
from sktime.forecasting.compose._reduce import RecursiveReductionForecaster, DirectReductionForecaster

# configure the pooling scehe to be used
POOLING = "local"

# generate  dummy data with a simple structure
# X: N_rows x N_columns (positive incremental integers)
# y: N_rows (negative decremental integers)
N_rows = 10
N_columns = 1
X_list = [list(range((v + 1) * 100 + 1, (v + 1) * 100 + N_rows + 1)) for v in range(N_columns)]
index = pd.period_range(periods=N_rows, freq="M", start="2021-01")
X = pd.DataFrame(X_list).T
X.columns = [f"column_{i+1}" for i in range(N_columns)]
X.index = index
y = pd.DataFrame(list(range(-1, -N_rows - 1, -1)), index=index, columns=["T"])


# special sklearn model in order to inspect inputs into model methods
class DummyRegressorWithPrintout(DummyRegressor):
   """Sklearn dummy model that would print out input data before fitting and predicting"""
    def fit(self, X, y):
        print(f"## fit ")
        print(f"### ---------------- X -------------------")
        print(f"```\n{X}\n```")
        print(f"### ---------------- y -------------------")
        print(f"```\n{y}\n```")
        m = super().fit(X, y)
        return m

    def predict(self, X):
        print(f"## predict")
        print(f"### ---------------- X -------------------")
        print(f"```\n{X}\n```")
        return super().predict(X)
estimator = DummyRegressorWithPrintout()

# setup a sktime reduction based on the sklearn model
forecasting_horizon = ForecastingHorizon([1, 2, 3], is_relative=True)
REDUCTION_SETTINGS = {
    "window_length": 2,
    "pooling": POOLING,
}
# old classes
recursive_old = RecursiveTabularRegressionForecaster(estimator, **REDUCTION_SETTINGS)
direct_old = DirectTabularRegressionForecaster(estimator, **REDUCTION_SETTINGS)
# new classes
recursive_new = RecursiveReductionForecaster(estimator, impute_method="bfill", **REDUCTION_SETTINGS)
direct_new = DirectReductionForecaster(estimator, impute_method="bfill", X_treatment="shifted", **REDUCTION_SETTINGS)

# loop over reductions: first train than predict on the whole dataset => no "future" X values in `predict`
for forecaster in [recursive_old, direct_old, recursive_new, direct_new]:
    print(f"# {forecaster.__class__.__name__}")
    try:
        forecaster = forecaster.fit(y, X, forecasting_horizon)
        preds = forecaster.predict(forecasting_horizon, X)
    except KeyError as err:
        print("> [!CAUTION]")
        print(f"> {err}")
```

**Comment by ericjb** [comment]
I submitted PR #7380 for speedup of _predict_out_of_sample for RecursiveReductionForecaster. See @fkiraly comment from July 2, 2023 in this issue for background.
